server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java

```diff
@@ -21,214 +21,170 @@
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
 
 import static org.elasticsearch.discovery.DiscoveryModule.DISCOVERY_HOSTS_PROVIDER_SETTING;
 import static org.elasticsearch.discovery.zen.SettingsBasedHostsProvider.DISCOVERY_ZEN_PING_UNICAST_HOSTS_SETTING;
 
 public class ClusterBootstrapService {
 
-    private static final Logger logger = LogManager.getLogger(ClusterBootstrapService.class);
-
     public static final Setting<List<String>> INITIAL_MASTER_NODES_SETTING =
         Setting.listSetting("cluster.initial_master_nodes", Collections.emptyList(), Function.identity(), Property.NodeScope);
 
     public static final Setting<TimeValue> UNCONFIGURED_BOOTSTRAP_TIMEOUT_SETTING =
         Setting.timeSetting("discovery.unconfigured_bootstrap_timeout",
             TimeValue.timeValueSeconds(3), TimeValue.timeValueMillis(1), Property.NodeScope);
 
+    private static final Logger logger = LogManager.getLogger(ClusterBootstrapService.class);
     private final List<String> initialMasterNodes;
-    @Nullable
+    @Nullable // null if discoveryIsConfigured()
     private final TimeValue unconfiguredBootstrapTimeout;
     private final TransportService transportService;
-    private volatile boolean running;
+    private final Supplier<Iterable<DiscoveryNode>> discoveredNodesSupplier;
+    private final Consumer<VotingConfiguration> votingConfigurationConsumer;
+    private final AtomicBoolean bootstrappingPermitted = new AtomicBoolean(true);
 
-    public ClusterBootstrapService(Settings settings, TransportService transportService) {
+    public ClusterBootstrapService(Settings settings, TransportService transportService,
+                                   Supplier<Iterable<DiscoveryNode>> discoveredNodesSupplier,
+                                   Consumer<VotingConfiguration> votingConfigurationConsumer) {
         initialMasterNodes = INITIAL_MASTER_NODES_SETTING.get(settings);
         unconfiguredBootstrapTimeout = discoveryIsConfigured(settings) ? null : UNCONFIGURED_BOOTSTRAP_TIMEOUT_SETTING.get(settings);
         this.transportService = transportService;
+        this.discoveredNodesSupplier = discoveredNodesSupplier;
+        this.votingConfigurationConsumer = votingConfigurationConsumer;
     }
 
     public static boolean discoveryIsConfigured(Settings settings) {
         return Stream.of(DISCOVERY_HOSTS_PROVIDER_SETTING, DISCOVERY_ZEN_PING_UNICAST_HOSTS_SETTING, INITIAL_MASTER_NODES_SETTING)
             .anyMatch(s -> s.exists(settings));
     }
 
-    public void start() {
-        assert running == false;
-        running = true;
+    void onFoundPeersUpdated() {
+        final Set<DiscoveryNode> nodes = getDiscoveredNodes();
+        if (transportService.getLocalNode().isMasterNode() && initialMasterNodes.isEmpty() == false
+            && nodes.stream().noneMatch(Coordinator::isZen1Node) && checkWaitRequirements(nodes)) {
```

Comment:
should we also have a check that we only start the bootstrap if we're not already bootstrapped?

Reply:
we potentially log `bootstrapping cancelled` at warning level even though bootstrapping is not required to happen

Reply:
... if the bootstrapping configuration is horribly broken. But yes, the message is misleading, I added a best-effort check in 8df9f666796176aa61e480572ecd0f6f6e443624.